rust/pattern/constraint/type_constraint.rs

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+use crate::{Constraint, TypeVariable, TypeVariableBuilder, UnboundVariable};
+use std::fmt;
+use std::fmt::{Display, Formatter};
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum TypeConstraint {
+    Label(LabelConstraint),
+    Sub(SubConstraint),
+    Relates(RelatesConstraint),
+    Plays(PlaysConstraint),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ScopedType {
+    scope: Option<String>,
+    name: String,
+}
```

Comment:
This is not used everywhere it should be used at the moment. Ideally any type label, scoped or unscoped, would be represented by this kind of structure.

Reply:
Nothing's stopping us, it's just low among the list of priorities. 

`Label` is right.